### ISSUE TITLE
Crash when using compsize on btrfs drives

### ISSUE THREAD

**Issue by Henri-J-Norden** [trapexit/mergerfs#1062]
**Describe the bug**

MergerFS crashes when calling [compsize](https://manpages.ubuntu.com/manpages/jammy/man8/compsize.8.html) on any directory in the pool (`compsize /mnt/raid/`). It does not crash when called on a file (`compsize /mnt/raid/file`), but compsize still fails and prints `All empty or still-delalloced files`.

Interestingly official btrfs-progs simply say `ERROR: not a btrfs filesystem: /mnt/raid` and refuse to do anything with the MergerFS pool... but it would be fantastic if compsize could actually work (i.e. MergerFS would forward the appropriate calls or something?)

**To Reproduce**

The issue is present even with just a single drive in the "pool". The drive has a btrfs FS, made with `mkfs.btrfs` default values and is mounted with `compress-force=zstd` (see fstab entries below).

1. Install btrfs-compsize
2. Mount MergerFS pool with a btrfs drive
3. Run compsize on any directory in the pool

**Expected behavior**

`compsize` should output the disk usage and compression ratio for the directory.

**System information:**

 - OS, kernel version: `Linux - 5.15.0-46-generic #49-Ubuntu SMP Thu Aug 4 18:03:25 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
 - mergerfs version: ` 2.33.5`
 - List of drives, filesystems, & sizes:
```
NAME     SIZE FSTYPE FSSIZE FSUSE% LABEL        MOUNTPOINTS
sdd      2.7T btrfs    2.7T    32% 2022-08#2 D4 /mnt/.data/4
```

 - mergerfs settings (fstab entries)
```
/mnt/.data/4 /mnt/raid fuse.mergerfs allow_other,use_ino,fsname=SnapRAID-MergerFS,moveonenospc=mspmfs,category.create=mspmfs,minfreespace=20G 0 0

/dev/disk/by-uuid/<uuid> /mnt/.data/4 auto nosuid,nodev,nofail,compress-force=zstd 0 0
```

**Attached files:**
Run on file (no crash, but doesn't work) `compsize /mnt/raid/backup/2022-08-29_root_ssd`
[file-mergerfs.strace.txt](https://github.com/trapexit/mergerfs/files/9482762/file-mergerfs.strace.txt)
[file-compsize.strace.txt](https://github.com/trapexit/mergerfs/files/9482763/file-compsize.strace.txt)

Run on directory (crash) `compsize /mnt/raid/backup/`
[dir-compsize.strace.txt](https://github.com/trapexit/mergerfs/files/9482758/dir-compsize.strace.txt)
[dir-mergerfs.strace.txt](https://github.com/trapexit/mergerfs/files/9482759/dir-mergerfs.strace.txt)

 - A strace of the application having a problem:
   - `strace -fvTtt -s 256 -o /tmp/app.strace.txt <cmd>`
   - `strace -fvTtt -s 256 -o /tmp/app.strace.txt -p <appPID>`
 - strace of mergerfs while app tried to do it's thing:
   - `strace -fvTtt -s 256 -p <mergerfsPID> -o /tmp/mergerfs.strace.txt`

**Comment by trapexit** [comment]
Interesting. Will take a look.

**Comment by trapexit** [comment]
As I suspected it's using an ioctl call and those are basically wide open in terms of form and behavior. I would probably need to special case this one. I try to pass through ioctl calls but it only works for basic ones.

**Comment by trapexit** [comment]
Interestingly... it doesn't crash for files or dirs. I think it might be coincidence. So long as there are files in the directory it looks like it just crawls the files and does the same call on them as individual files. And due to the complexity of the ioctl call it's causing it to crash... doing things it shouldn't wrt memory.

How would you see this thing working? It's making a direct call to the filesystem so I guess I would have to run the call on all branches and then aggregate them? 

This is a tough problem to solve. I either have to whitelist and expressly proxy specific calls known to be safe or blacklist those known to be bad. Or allow anything and when this kind of situation happens try to address the specific request. The ioctl interface is just a wide open mess and it's kinda dangerous for me to allow them generally as I have... as you can tell :) 

I'm surprised it's been so many years till someone hit this.

**Comment by chapmanjacobd** [comment]
yeah I also got "Transport endpoint is not connected" when I accidentally queried a mergerfs mount with a btrfs tool.

Here is the strace output for that btrfs tool on an actual btrfs disk since I don't want to have mergerfs crash again. I imagine the calls will be the same

[strace.txt](https://github.com/trapexit/mergerfs/files/10580022/strace.txt)

**Comment by trapexit** [comment]
Yeah. It's probably some complex ioctl. mergerfs tries to just pass through calls but due to the flexible nature of ioctl it can lead to  these crashes because the data isn't properly handled. Unfortunately, there just isn't a good solution. 

1) whitelist known good ioctl calls
2) blacklist known bad ioctl calls
3) white list known good ones and try to handle others based on user feedback when they try something and it fails.

Since btrfs calls are the topic here I could go look them all up and just return an error any time I see them. That would be the least impactful and easiest change. I was planning on doing this for the next release to close this out anyway.

**Comment by trapexit** [comment]
"fixed" in #1137 

If there are calls that people want to try to use through mergerfs I'll need to investigate them one at a time to see if it is possible.
